thriftbp/server_transport.go

```diff
@@ -0,0 +1,48 @@
+package thriftbp
+
+import (
+	"github.com/apache/thrift/lib/go/thrift"
+	"github.com/go-kit/kit/metrics"
+	"github.com/reddit/baseplate.go/metricsbp"
+)
+
+const meterNameTransportConnCounter = "thrift.connections"
+
+type CountedTServerTransport struct {
+	thrift.TServerTransport
+}
+
+func (m *CountedTServerTransport) Accept() (thrift.TTransport, error) {
+
+	transport, err := m.TServerTransport.Accept()
+	if err != nil {
+		return nil, err
+	}
+
+	return NewCountedTTransport(transport), nil
+}
+
+type CountedTTransport struct {
+	thrift.TTransport
+	counter metrics.Counter
+}
+
+func NewCountedTTransport(transport thrift.TTransport) thrift.TTransport {
+	return &CountedTTransport{
+		TTransport: transport,
+		counter:    metricsbp.M.Counter(meterNameTransportConnCounter),
+	}
+}
+
+func (m *CountedTTransport) Close() error {
+	m.counter.Add(-1)
```

Comment:
`TTransport.Close` can be called multiple times (calling it on an already closed transport will just be no-op or return an "already closed" error, can't remember exactly which), so just blindly doing -1 will likely give you inaccurate numbers.

There's `TTransport.IsOpen` you can use, but that will likely also give you inaccurate number, because it can return false without explicitly calling `Close` (for example, the other end closed the connection).

I think the closest we can get to a reliable way to implement it is to add some atomic value in `CountedTTransport`, so that we make sure that we only do -1 once and the future `Close` calls will not try to -1 again.

Reply:
I would put a `closeOnce sync.Once` in `CountedTTransport` and do

```
m.closeOnce.Do(func() {
  m.counter.Add(-1)
})
```

Reply:
yea once should work

Reply:
👍

Reply:
another concern I have is that we are not really using true counters with statsd metrics. our telegraf configuration will reset every counter every minute, which means there could be a scenario that in a minute window we report more -1s than +1s, and causing telegraf to report a negative counter to the upstream, and that could cause problems.

we won't have that problem on prometheus (as we use true counters), but for statsd it probably makes more sense to use runtime gauges, so we report the per-pod gauge, instead of a global counter.

Reply:
So we use a counter for the inflight metrics of our rate limiter queue and havent run into this yet. Considering that we plan to move to prometheus and this is a new API we can keep it as is and verify in the beta phase.